### DISCUSSION TITLE
next-firebase-auth vs NextAuth, what is the difference?

### DISCUSSION THREAD

**Discussion by abusada** [gladly-team/next-firebase-auth#522] [Q&A] [answered]
What is the difference between next-firebase-auth, and [NextAuth](https://next-auth.js.org/adapters/firebase) using the firebase adapter?

**Comment by kmjennison** [comment, marked as answer]
NextAuth.js and `next-firebase-auth` both get the job done but have a few differences.

**NextAuth.js:**
* Does not use Firebase's authentication. It uses Google OAuth (or another identity provider) and provides a _Firestore_ adapter to store session info. In other words, users will not appear in the Firebase authentication console; e.g., https://console.firebase.google.com/project/example-project/authentication/users
* Is undoubtedly more flexible: there are lots of authentication providers and database adapters
* Requires the developer to set up a session database for email sign-in
* Does not directly provide access to the user's Firebase ID token in server-side rendering
* Supports multiple identity providers for the same user _only if_ you use a session database. Linked accounts are managed by NextAuth.js.
* Includes some features `next-firebase-auth` does not support, such as tab syncing
* Is a little more complicated/technical to set up and manage (subjective assessment)

**`next-firebase-auth`:**
* Uses Firebase authentication and thus supports related Firebase authentication tooling (like FirebaseUI)
* Is limited to Firebase's authentication functionality and does not provide custom server-side session storage
* Does not require the developer to set up a session database
* Provides access to the user's Firebase ID token in server-side rendering
* Supports multiple identity providers for the same user. Linked accounts are managed by by Firebase's authentication logic.
* Is a little easier to set up and manage (subjective assessment)

**Both**:
* Support server-side rendering
* Support auth-protected routes
* Use signed, secure, server-only cookies by default

**Which should I use?**

This will depend on your app's needs and your preferences.

NextAuth.js is an excellent choice if you're looking for a generalized authentication solution. It provides more platform independence and flexibility. At the time of writing, the project seems well-supported and has a good community of developers. It would probably be my choice for a new, non-hobbyist project that doesn't plan to utilize the Firebase platform too much.

`next-firebase-auth` is an excellent choice if you want a more hands-off authentication backend or if you'll be relying on the Firebase platform elsewhere in your stack. In particular, it's easier to get started if you need to support email sign-in. It is, of course, more reliant on the Firebase platform. I'd choose it for a project I wanted to launch quickly and would use Firebase platform features.

**Disclaimer**

FYI, I'm the maintainer of `next-firebase-auth`.

**Thoughts from others?**

Others evaluating Next.js auth library options—please chime in below!

**Reply by abusada** [reply]
Thank you for the very elaborate and clear explanation, I think this should be part of the docs somewhere, don't you think?

**Reply by kmjennison** [reply]
Good suggestion. Added a link here from the README:
https://github.com/gladly-team/next-firebase-auth/pull/524

**Reply by sraka1** [reply]
Great writeup @kmjennison, big thanks. I have an additional question, btw: any specific reason why you opted to develop a separate library rather than add Firebase as a identity provider to NextAuth.js?

**Reply by kmjennison** [reply]
@sraka1 Initially, work deadlines made this the easier path forward. Now, it's possible a new NextAuth.js provider could handle the same functionality, but I haven't looked into it enough to say for sure.

**Reply by elbro** [reply]
is this updated to work with NextJs 13 and would ths comparison still hold true?

**Reply by kmjennison** [reply]
@elbro See #568 -- this works with Next.js 13 but not yet for the new app directory.

**Reply by supto09** [reply]
Any ETA when we can use this awesome library with app directory? Can't wait to use both of them..

**Reply by marceloverdijk** [reply]
@kmjennison I assume this still does not work with the new app directory? Maybe a good idea to put a disclaimer on the main page?

**Reply by kmjennison** [reply]
@marceloverdijk Good idea. Added: https://github.com/gladly-team/next-firebase-auth/commit/d51bf07eecf727ef3df45587e4008551b0cb4803
